src/Components/Aspire.Azure.Npgsql.EntityFrameworkCore.PostgreSQL/AspireAzureEFPostgreSqlExtensions.cs

```diff
@@ -0,0 +1,171 @@
+        destination.DisableTracing = source.DisableTracing;
+        destination.DisableRetry = source.DisableRetry;
+        destination.CommandTimeout = source.CommandTimeout;
+    }
+
+    private static void ConfigureDbContextOptionsBuilder(AzureNpgsqlEntityFrameworkCorePostgreSQLSettings settings, DbContextOptionsBuilder dbContextOptionsBuilder)
+    {
+        // The connection string requires the username to be provided. Since it will depend on the Managed Identity that is used
+        // we attempt to get the username from the access token.
+
+        var credential = settings.Credential ?? new DefaultAzureCredential();
+
+#pragma warning disable EF1001 // Internal EF Core API usage.
+
+        // Get the connection string from the Npgsql options extension in case it was set using UseNpgsql(connStr) and Enrich()
+        var extensionsConnectionString = dbContextOptionsBuilder.Options.GetExtension<NpgsqlOptionsExtension>()?.ConnectionString;
+
+#pragma warning restore EF1001 // Internal EF Core API usage.
+
+        var dataSourceBuilder = new NpgsqlDataSourceBuilder(settings.ConnectionString ?? extensionsConnectionString);
+
+        if (string.IsNullOrEmpty(dataSourceBuilder.ConnectionStringBuilder.Username))
+        {
+            // Ensure to use the management scope, so the token contains user names for all managed identity types - e.g. user and service principal
+            var token = credential.GetToken(ManagedIdentityTokenCredentialHelpers.ManagementTokenRequestContext, default);
+
+            if (ManagedIdentityTokenCredentialHelpers.TryGetUsernameFromToken(token.Token, out var username))
+            {
+                dataSourceBuilder.ConnectionStringBuilder.Username = username;
+            }
+            else
+            {
+                // Otherwise check using the PostgresSql scope
+                token = credential.GetToken(ManagedIdentityTokenCredentialHelpers.DatabaseForPostgresSqlTokenRequestContext, default);
+
+                if (ManagedIdentityTokenCredentialHelpers.TryGetUsernameFromToken(token.Token, out username))
+                {
+                    dataSourceBuilder.ConnectionStringBuilder.Username = username;
+                }
+            }
+
+            // If we still don't have a username, we let Npgsql handle the error when trying to connect.
+        }
+
+        if (string.IsNullOrEmpty(dataSourceBuilder.ConnectionStringBuilder.Password))
+        {
+            dataSourceBuilder.UsePasswordProvider(
+                passwordProvider: _ => credential.GetToken(ManagedIdentityTokenCredentialHelpers.DatabaseForPostgresSqlTokenRequestContext, default).Token,
+                passwordProviderAsync: async (_, ct) => (await credential.GetTokenAsync(ManagedIdentityTokenCredentialHelpers.DatabaseForPostgresSqlTokenRequestContext, default).ConfigureAwait(false)).Token
+            );
+
+            dbContextOptionsBuilder.UseNpgsql(dataSourceBuilder.Build());
```

Comment:
This might be a dumb scenario, but what happens if someone doesn't have a UserName, but does have a Password? We wouldn't be setting the `dataSourceBuilder` in that case. But maybe we don't care.